### ISSUE TITLE
strings templated using a lookup in a loop shouldn't be cached and re-used

### ISSUE THREAD

**Issue by qiwangrh** [ansible/ansible#82955]
### Summary

I was trying to run ansible playbook with the following tasks to frequently monitor if a host's IP was changed or not. Once the IP was changed, I expected the playbook exited the retries loop. The actual result that I got is: the playbook couldn't detect the ip address' change.
```
- name: Get initial deployment engine ip address
  ansible.builtin.set_fact:
    initial_deployment_engine_ip: "{{ lookup('community.general.dig', deployment_engine_fqdn) }}"

- name: Print initial deployment engine ip address
  ansible.builtin.debug:
    msg: "{{ initial_deployment_engine_ip }}"

- name: Check every 2 mins if the deployment completes
  ansible.builtin.set_fact:
    deployment_engine_ip: "{{ lookup('community.general.dig', deployment_engine_fqdn) }}"
  until: deployment_engine_ip != initial_deployment_engine_ip 
  retries: 60
  delay: 12 
```
It might be a bug in Ansible. The issue might be the template string gets rendered once, but the set_fact probably does fire off multiple times (with the now constant value from the templated string). Strings templated using a lookup in a loop shouldn't be cached and re-used.

### Issue Type

Feature Request

### Component Name

set_fact

### Ansible Version

```console
$ ansible --version
ansible [core 2.16.2]
  config file = /home/qwa/.ansible.cfg
  configured module search path = ['/home/qwa/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/qwa/.local/lib/python3.12/site-packages/ansible
  ansible collection location = /home/qwa/.ansible/collections:/usr/share/ansible/collections
  executable location = /home/qwa/.local/bin/ansible
  python version = 3.12.0 (main, Oct  2 2023, 00:00:00) [GCC 13.2.1 20230918 (Red Hat 13.2.1-3)] (/usr/bin/python3)
  jinja version = 3.1.2
  libyaml = True
```

### Configuration

```console
# if using a version older than ansible-core 2.12 you should omit the '-t all'
$ ansible-config dump --only-changed -t all
CONFIG_FILE() = /home/qwa/.ansible.cfg
```

### OS / Environment

Fedora

### Steps to Reproduce

You can use the following way to reproduce the issue.
```
---
- hosts: localhost
  tasks:
    - name: Set original fact
      ansible.builtin.set_fact:
        original_value: "{{ lookup('ansible.builtin.file', '/home/jrb/experiments/set_fact/foo.txt') }}"

    - name: Show the value
      ansible.builtin.debug:
        var: original_value

    - name: Wait until the value changes
      ansible.builtin.set_fact:
        new_value: "{{ lookup('ansible.builtin.file', '/home/jrb/experiments/set_fact/foo.txt') }}"
      until: new_value != original_value
      retries: 3
      delay: 10
```

### Expected Results

I expect the above playbook would exit the retries loop when it detected the changes in the foo.txt file.

### Actual Results

```console
I did an echo "Bar" > foo.txt after the first iteration of the task, so it ought to have changed.

[jrb@unknown8c1645fec0e5 set_fact]$ ansible-playbook -vvvv playbook.yml
ansible-playbook [core 2.16.3]
  config file = /home/jrb/experiments/set_fact/ansible.cfg
  configured module search path = ['/home/jrb/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/jrb/.local/lib/python3.10/site-packages/ansible
  ansible collection location = /home/jrb/experiments/set_fact/collections
  executable location = /home/jrb/.local/bin/ansible-playbook
  python version = 3.10.11 (main, Apr  5 2023, 00:00:00) [GCC 12.2.1 20221121 (Red Hat 12.2.1-4)] (/usr/bin/python3)
  jinja version = 3.0.3
  libyaml = True
Using /home/jrb/experiments/set_fact/ansible.cfg as config file
setting up inventory plugins
Loading collection ansible.builtin from 
host_list declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
Skipping due to inventory source not existing or not being readable by the current user
script declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
auto declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
Skipping due to inventory source not existing or not being readable by the current user
yaml declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
Skipping due to inventory source not existing or not being readable by the current user
ini declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
Skipping due to inventory source not existing or not being readable by the current user
toml declined parsing /etc/ansible/hosts as it did not pass its verify_file() method
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that
the implicit localhost does not match 'all'
Loading callback plugin default of type stdout, v2.0 from /home/jrb/.local/lib/python3.10/site-packages/ansible/plugins/callback/default.py
Skipping callback 'default', as we already have a stdout callback.
Skipping callback 'minimal', as we already have a stdout callback.
Skipping callback 'oneline', as we already have a stdout callback.PLAYBOOK: playbook.yml *********************************************************
Positional arguments: playbook.yml
verbosity: 4
connection: ssh
become_method: sudo
tags: ('all',)
inventory: ('/etc/ansible/hosts',)
forks: 5
1 plays in playbook.ymlPLAY [localhost] ***************************************************************TASK [Gathering Facts] *********************************************************
task path: /home/jrb/experiments/set_fact/playbook.yml:2
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: jrb
<127.0.0.1> EXEC /bin/sh -c 'echo ~jrb && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /home/jrb/.ansible/tmp `"&& mkdir "` echo /home/jrb/.ansible/tmp/ansible-tmp-1712002162.820464-19670-61381277522406 `" && echo ansible-tmp-1712002162.820464-19670-61381277522406="` echo /home/jrb/.ansible/tmp/ansible-tmp-1712002162.820464-19670-61381277522406 `" ) && sleep 0'
Using module file /home/jrb/.local/lib/python3.10/site-packages/ansible/modules/setup.py
<127.0.0.1> PUT /home/jrb/.ansible/tmp/ansible-local-19667_pcff9o9/tmpykyc9ynl TO /home/jrb/.ansible/tmp/ansible-tmp-1712002162.820464-19670-61381277522406/AnsiballZ_setup.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /home/jrb/.ansible/tmp/ansible-tmp-1712002162.820464-19670-61381277522406/ /home/jrb/.ansible/tmp/ansible-tmp-1712002162.820464-19670-61381277522406/AnsiballZ_setup.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python3 /home/jrb/.ansible/tmp/ansible-tmp-1712002162.820464-19670-61381277522406/AnsiballZ_setup.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /home/jrb/.ansible/tmp/ansible-tmp-1712002162.820464-19670-61381277522406/ > /dev/null 2>&1 && sleep 0'
ok: [localhost]TASK [Set original fact] *******************************************************
task path: /home/jrb/experiments/set_fact/playbook.yml:4
File lookup using /home/jrb/experiments/set_fact/foo.txt as file
ok: [localhost] => {
    "ansible_facts": {
        "original_value": "Foo"
    },
    "changed": false
}TASK [Show the value] **********************************************************
task path: /home/jrb/experiments/set_fact/playbook.yml:8
ok: [localhost] => {
    "original_value": "Foo"
}TASK [Wait until the value changes] ********************************************
task path: /home/jrb/experiments/set_fact/playbook.yml:12
File lookup using /home/jrb/experiments/set_fact/foo.txt as file
FAILED - RETRYING: [localhost]: Wait until the value changes (3 retries left).Result was: {
    "ansible_facts": {
        "new_value": "Foo"
    },
    "attempts": 1,
    "changed": false,
    "retries": 4
}
FAILED - RETRYING: [localhost]: Wait until the value changes (2 retries left).Result was: {
    "ansible_facts": {
        "new_value": "Foo"
    },
    "attempts": 2,
    "changed": false,
    "retries": 4
}
FAILED - RETRYING: [localhost]: Wait until the value changes (1 retries left).Result was: {
    "ansible_facts": {
        "new_value": "Foo"
    },
    "attempts": 3,
    "changed": false,
    "retries": 4
}
fatal: [localhost]: FAILED! => {
    "ansible_facts": {
        "new_value": "Foo"
    },
    "attempts": 3,
    "changed": false
}PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   
```
The ansible_facts value is output 3 times, indicating to me that set_fact is firing off all 3 times, but the log message "File lookup using /home/jrb/experiments/set_fact/foo.txt as file" message was output only once on the first loop indicating that the value is reused.
```


### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

**Comment by ansibot** [comment]
Files identified in the description:

* [`lib/ansible/plugins/action/set_fact.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/set_fact.py)
* [`lib/ansible/modules/set_fact.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/set_fact.py)

If these files are incorrect, please update the `component name` section of the description or use the [component](https://github.com/ansible/ansibotmini#commands) bot command.

**Comment by sivel** [comment]
The cause of this beahvior is that the task defintion is only `post_validated` 1 time, which is when the task args are templated.

So before the `set_fact` task action plugin actually even executes, the task has been finalized with a static value, so it is not possible for `until` to expect a different result.  Allowing this would have far reaching consequences. 

Ultimately this is a duplicate of https://github.com/ansible/ansible/issues/78266

**Comment by qiwangrh** [comment]
@sivel , thanks for your analysis. I saw [#78266](https://github.com/ansible/ansible/issues/78266) was already closed. 
Ansible is new to me. I am not sure what else I need to check for the reported issue. Could you suggest?

**Comment by qiwangrh** [comment]
Thanks @sivel , I had a workaround for my playbook. Here I am trying to understand if the reported issue is a problem which needs a fix or not. If it is not an issue, I will get it closed.

**Comment by flowerysong** [comment]
You can also move the lookup so that it's part of the `until` condition, which *does* get evaluated each time.

```yaml
- hosts: localhost
  vars:
    new_value: "{{ lookup('ansible.builtin.file', '/var/tmp/foo.txt') }}"
  tasks:
    - name: Set original fact
      ansible.builtin.set_fact:
        original_value: "{{ new_value }}"

    - name: Wait until the value changes
      ansible.builtin.debug:
        msg: Waiting for value to change from {{ original_value }}
      until: new_value != original_value
      retries: 3
      delay: 10

    - name: Display new value
      ansible.builtin.debug:
        msg: "{{ new_value }}"
```

```
PLAY [localhost] ***************************************************************

TASK [Set original fact] *******************************************************
ok: [localhost]

TASK [Wait until the value changes] ********************************************
FAILED - RETRYING: [localhost]: Wait until the value changes (3 retries left).Result was:
    attempts: 1
    changed: false
    msg: Waiting for value to change from foobar
    retries: 4
ok: [localhost] =>
    msg: Waiting for value to change from foobar

TASK [Display new value] *******************************************************
ok: [localhost] =>
    msg: foo
```

**Comment by qiwangrh** [comment]
Thanks @flowerysong, I used the same workaround as you mentioned. It worked.

**Comment by s-hertel** [comment]
@qiwangrh I reclassified this as a feature (we can't change the behavior without breaking content since it is working as designed), but how different keywords interact with different stages of task execution is definitely something we need to document better. There's an open issue for that here: https://github.com/ansible/ansible-documentation/issues/1295.

I just merged a loop_control break_when feature for 2.18 (#62151) that can be used like this:

```
- ansible.builtin.set_fact:
    deployment_engine_ip: "{{ lookup('community.general.dig', deployment_engine_fqdn) }}"
  loop: "{{ range(0, 60) }}"
  loop_control:
    pause: 12
    break_when: deployment_engine_ip != initial_deployment_engine_ip
```

Does this seem like a good resolution?

**Comment by mkrizek** [comment]
I am going to go ahead and close this issue as there are workarounds as well as the `break_when` functionality shown above.

If you have further questions please stop by the Ansible Forum:

* [Ansible Forum](https://forum.ansible.com)

See this page for a complete and up to date list of communication channels and their purposes:

* <https://docs.ansible.com/ansible/latest/community/communication.html>
